utilities/source/index.mts

```diff
@@ -266,33 +248,92 @@ export function intern<
           "undefined",
         ].includes(typeof innerValue) ||
         innerValue === null ||
-        (innerValue as any)[internSymbol] === true
+        intern.isInterned(innerValue)
       )
     )
       throw new Error(
         `Failed to intern value because of non-interned inner value.`,
       );
-  const key = Symbol();
-  (value as any)[internSymbol] = true;
-  Object.freeze(value);
-  intern.pool[type].set(key, new WeakRef(value as any));
-  intern.finalizationRegistry.register(value, { type, key });
-  return value as any;
-}
 
-export const internSymbol = Symbol("intern");
+  const entries = Array.isArray(value)
+    ? value.entries()
+    : Object.entries(value).sort(([aKey], [bKey]) => aKey.localeCompare(bKey));
+
+  // Find leaf node, creating intermediary nodes as necessary
+  let node = intern._rootInternNode;
+  for (const [key, innerValue] of entries) {
+    if (node.children === undefined) node.children = new Map();
+    if (!node.children.has(key)) node.children.set(key, new Map());
+    const valueMap = node.children.get(key)!;
+    if (!valueMap.has(innerValue))
+      valueMap.set(innerValue, { parent: node, key: key, value: innerValue });
+    node = valueMap.get(innerValue)!;
+  }
+
+  // Special case empty object
+  if (node.root) {
+    return (Array.isArray(value) ? nullTuple : nullRecord) as any;
+  }
+
+  // If we already have a value, return it
+  if (node.internedValue !== undefined) return node.internedValue.deref()!;
+
+  // Otherwise intern the value and cache it
+  intern._markInterned(value);
+  node.internedValue = new WeakRef(value);
+  intern._finalizationRegistry.register(value, node);
+  return value as Interned<T>;
+}
 
-intern.pool = {
-  tuple: new Map<Symbol, WeakRef<Intern<InternInnerValue[]>>>(),
-  record: new Map<
-    Symbol,
-    WeakRef<Intern<{ [key: string]: InternInnerValue }>>
-  >(),
+const internSymbol = Symbol("intern");
+intern._markInterned = (value: any) => {
+  Object.defineProperty(value, internSymbol, {
+    enumerable: false,
+    value: true,
+  });
+  Object.freeze(value);
 };
 
-intern.finalizationRegistry = new FinalizationRegistry<{
-  type: "tuple" | "record";
-  key: Symbol;
-}>(({ type, key }) => {
-  intern.pool[type].delete(key);
+intern.isInterned = (value: any): boolean =>
+  (value as any)[internSymbol] === true;
+
+const nullTuple: any[] = [];
+intern._markInterned(nullTuple);
+const nullRecord = {};
+intern._markInterned(nullRecord);
+
+type InternNode =
+  | {
+      /** The intermediary key for this node ie `node.key = node.parent.get(node.key).get(node.value).key` */
+      key: InternInnerKey;
+      /** The intermediary value for this node ie `node.value = node.parent.get(node.key).get(node.value).value` */
+      value: InternInnerValue;
```

Comment:
Some points on naming here:
* Node feels a bit generic
* I'm not sure I love `inner`. For records `key=property_name, value: property_value`, For tuples key is the index in the array and property value is the value at that index. We never have a conflict with the two because `typeof Object.keys({ 1: 1 })[0] === 'string'`
* `Intern` prefix is just namespacing, if this was in a separate file we could lose the prefix